<table dir="ltr" width="100%"><thead><tr><th width="50%" dir="ltr" lang="en" align="left">English</th><th width="50%" dir="rtl" lang="ar" align="right">العربية</th></tr></thead><tbody><tr><td width="50%" dir="ltr" lang="en" align="left" valign="top"><h1>Lab 01 · Real Delta Bronze</h1><p>SDA-DSC-214 · Meaad Al-Marri</p></td><td width="50%" dir="rtl" lang="ar" align="right" valign="top"><h1>اللاب 01 · Bronze فعلية باستخدام Delta</h1><p>SDA-DSC-214 · ميعاد المري</p></td></tr></tbody></table>

<table dir="ltr" width="100%"><tr><td width="50%" valign="top" dir="ltr" lang="en" align="left"><h2>Goal and recorded execution</h2><p>This section preserves the same source cells and actual outputs from the verified <a href="../../DAY01_STUDENT.ipynb">combined Day 1 notebook</a>. It is an alternative section of the same lab, not an extra assignment or a claim of a separate kernel run. Follow <a href="../../labs/lab01/WALKTHROUGH.md">the walkthrough</a> and <a href="../../STATUS.md">the execution record</a>. The scan section requires the successful Bronze workspace from Lab 01.</p></td><td width="50%" valign="top" dir="rtl" lang="ar" align="right"><h2>الهدف والتنفيذ المسجل</h2><p>يحفظ هذا المقطع الخلايا البرمجية نفسها ومخرجاتها الفعلية من <a href="../../DAY01_STUDENT.ipynb">دفتر اليوم الأول الموحد المتحقق</a>. هو مقطع بديل من اللاب نفسه، وليس تكليفًا إضافيًا أو ادعاء تشغيل مستقل بنواة أخرى. اتبع <a href="../../labs/lab01/WALKTHROUGH.md">الشرح</a> و<a href="../../STATUS.md">سجل التنفيذ</a>. يتطلب مقطع القياس مساحة Bronze الناجحة من اللاب 01.</p></td></tr></table>

<table dir="ltr" width="100%"><thead><tr><th width="50%" dir="ltr" lang="en" align="left">English</th><th width="50%" dir="rtl" lang="ar" align="right">العربية</th></tr></thead><tbody><tr><td width="50%" dir="ltr" lang="en" align="left" valign="top"><h2>Setup · report the actual environment</h2><p>Only inspect dependencies here. This cell does not install packages or start Spark.</p></td><td width="50%" dir="rtl" lang="ar" align="right" valign="top"><h2>الإعداد · اعرض البيئة الفعلية</h2><p>افحص المتطلبات فقط هنا. لا تثبت هذه الخلية حزمًا ولا تبدأ Spark.</p></td></tr></tbody></table>

In [6]:
from pathlib import Path
import sys, json
ROOT = next((p for p in (Path.cwd(), *Path.cwd().parents)
             if (p / "course.json").is_file() and (p / "src/masar").is_dir()), None)
if ROOT is None:
    raise FileNotFoundError("Open the notebook from within the complete course repository")
sys.path.insert(0, str(ROOT / "src"))
SOURCE = ROOT / "data/masar-small-v1"
from masar.runtime import inspect_environment, require_environment, start_spark
print(json.dumps(inspect_environment(), indent=2))

{
  "scope": "DEPENDENCY_PREFLIGHT_ONLY",
  "python": "3.11.16",
  "java": "openjdk version \"17.0.20.1\" 2026-08-18",
  "java_major": 17,
  "packages": {
    "pyspark": {
      "required": "3.5.8",
      "observed": "3.5.8"
    },
    "delta-spark": {
      "required": "3.3.2",
      "observed": "3.3.2"
    },
    "py4j": {
      "required": "0.10.9.9",
      "observed": "0.10.9.9"
    }
  },
  "status": "DEPENDENCIES_PRESENT_ENGINE_NOT_TESTED",
  "issues": [],
  "engine_executed": false
}


<table dir="ltr" width="100%"><thead><tr><th width="50%" dir="ltr" lang="en" align="left">English</th><th width="50%" dir="rtl" lang="ar" align="right">العربية</th></tr></thead><tbody><tr><td width="50%" dir="ltr" lang="en" align="left" valign="top"><h2>Start a clean, bounded run</h2><p>Fail clearly if dependencies are missing; never switch engines silently. Existing work is retained.</p></td><td width="50%" dir="rtl" lang="ar" align="right" valign="top"><h2>ابدأ تشغيلًا جديدًا محدود النطاق</h2><p>توقف بوضوح عند نقص المتطلبات دون تبديل المحرك بصمت. تبقى الأعمال السابقة محفوظة.</p></td></tr></tbody></table>

In [7]:
require_environment()
from masar.workspace import new_workspace, require_fixed_dataset, write_json, workspace_path
require_fixed_dataset(SOURCE)
WORK = new_workspace(ROOT, "day01_bronze")
print("Workspace:", WORK.relative_to(ROOT))

Workspace: outputs/day01_bronze_i4xhsr5r


<table dir="ltr" width="100%"><thead><tr><th width="50%" dir="ltr" lang="en" align="left">English</th><th width="50%" dir="rtl" lang="ar" align="right">العربية</th></tr></thead><tbody><tr><td width="50%" dir="ltr" lang="en" align="left" valign="top"><h2>Read one source</h2><p>CSV strings remain unchanged. The count action starts computation.</p></td><td width="50%" dir="rtl" lang="ar" align="right" valign="top"><h2>اقرأ مصدرًا واحدًا</h2><p>تبقى نصوص CSV كما هي. يطلق فعل العد الحساب الفعلي.</p></td></tr></tbody></table>

In [8]:
from masar.bronze import raw_frame, ingest_feed, verify_bronze
# Functions are imported without starting Spark. The next cell executes the lab.
print("Bronze source code loaded; engine has not started yet.")

Bronze source code loaded; engine has not started yet.


<table dir="ltr" width="100%"><thead><tr><th width="50%" dir="ltr" lang="en" align="left">English</th><th width="50%" dir="rtl" lang="ar" align="right">العربية</th></tr></thead><tbody><tr><td width="50%" dir="ltr" lang="en" align="left" valign="top"><h2>Load, replay, verify and retain evidence</h2><p>The <code>try/finally</code> always stops the session after this block, even on failure. Reuse of a committed batch id is rejected. See <a href="../../src/masar/bronze.py">shared source</a>; this is not a placeholder or mock engine.</p></td><td width="50%" dir="rtl" lang="ar" align="right" valign="top"><h2>استوعب وأعد الوصول وتحقق واحفظ الدليل</h2><p>توقف <code>try/finally</code> الجلسة بعد هذه الكتلة حتى عند الفشل. تُرفض إعادة استخدام معرف دفعة محفوظ. راجع <a href="../../src/masar/bronze.py">المصدر المشترك</a>؛ ليست هذه محاكاة للمحرك أو موضعًا فارغًا.</p></td></tr></tbody></table>

In [9]:
from masar.workspace import record_bronze_success
spark = start_spark(WORK)
try:
    print("Spark:", spark.version)
    raw_trips = raw_frame(spark, SOURCE, "trips")
    raw_trips.printSchema()
    raw_trips.show(3, truncate=False)
    print("Source trips:", raw_trips.count())
    for feed in ("trips", "drivers", "gps_events"):
        print(ingest_feed(spark, SOURCE, WORK, feed, "base_001"))
    print(ingest_feed(spark, SOURCE, WORK, "trips", "replay_002"))
    report = verify_bronze(spark, SOURCE, WORK)
    record_bronze_success(ROOT, WORK)
    print(json.dumps({"counts": report["counts"], "checks": report["checks"]}, indent=2))
    print("Retained:", (WORK / "reports/bronze.json").relative_to(ROOT))
finally:
    spark.stop()

https://repo.maven.apache.org/maven2 added as a remote repository with the name: repo-1


:: loading settings :: url = jar:file:/opt/hostedtoolcache/Python/3.11.16/x64/lib/python3.11/site-packages/pyspark/jars/ivy-2.5.1.jar!/org/apache/ivy/core/settings/ivysettings.xml


Ivy Default Cache set to: /home/runner/.ivy2/cache
The jars for the packages stored in: /home/runner/.ivy2/jars
io.delta#delta-spark_2.12 added as a dependency
:: resolving dependencies :: org.apache.spark#spark-submit-parent-e718d787-4d49-46c2-bef0-46b4de1c78a1;1.0
	confs: [default]


	found io.delta#delta-spark_2.12;3.3.2 in central
	found io.delta#delta-storage;3.3.2 in central


	found org.antlr#antlr4-runtime;4.9.3 in central
downloading https://repo1.maven.org/maven2/io/delta/delta-spark_2.12/3.3.2/delta-spark_2.12-3.3.2.jar ...


	[SUCCESSFUL ] io.delta#delta-spark_2.12;3.3.2!delta-spark_2.12.jar (247ms)
downloading https://repo1.maven.org/maven2/io/delta/delta-storage/3.3.2/delta-storage-3.3.2.jar ...
	[SUCCESSFUL ] io.delta#delta-storage;3.3.2!delta-storage.jar (55ms)
downloading https://repo1.maven.org/maven2/org/antlr/antlr4-runtime/4.9.3/antlr4-runtime-4.9.3.jar ...
	[SUCCESSFUL ] org.antlr#antlr4-runtime;4.9.3!antlr4-runtime.jar (63ms)
:: resolution report :: resolve 2182ms :: artifacts dl 371ms
	:: modules in use:
	io.delta#delta-spark_2.12;3.3.2 from central in [default]
	io.delta#delta-storage;3.3.2 from central in [default]
	org.antlr#antlr4-runtime;4.9.3 from central in [default]
	---------------------------------------------------------------------
	|                  |            modules            ||   artifacts   |
	|       conf       | number| search|dwnlded|evicted|| number|dwnlded|
	---------------------------------------------------------------------
	|      default     |   3   |   3   |   3 

26/09/10 12:50:38 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable


Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).


Spark: 3.5.8


root
 |-- trip_id: string (nullable = true)
 |-- driver_id: string (nullable = true)
 |-- city: string (nullable = true)
 |-- start_ts: string (nullable = true)
 |-- end_ts: string (nullable = true)
 |-- fare_sar: string (nullable = true)
 |-- distance_km: string (nullable = true)



+---------+---------+------+-------------------------+-------------------------+--------+-----------+
|trip_id  |driver_id|city  |start_ts                 |end_ts                   |fare_sar|distance_km|
+---------+---------+------+-------------------------+-------------------------+--------+-----------+
|SYN_T0001|SYN_D001 |Riyadh|2026-06-01T06:00:00+03:00|2026-06-01T06:08:00+03:00|18.00   |3.50       |
|SYN_T0002|SYN_D002 |Riyadh|2026-06-01T08:00:00+03:00|2026-06-01T08:11:00+03:00|19.25   |3.85       |
|SYN_T0003|SYN_D001 |Riyadh|2026-06-01T10:00:00+03:00|2026-06-01T10:14:00+03:00|20.50   |4.20       |
+---------+---------+------+-------------------------+-------------------------+--------+-----------+
only showing top 3 rows



Source trips: 72


{'feed': 'trips', 'batch_id': 'base_001', 'appended_rows': 72, 'total_rows': 72, 'version': 0}


{'feed': 'drivers', 'batch_id': 'base_001', 'appended_rows': 6, 'total_rows': 6, 'version': 0}


{'feed': 'gps_events', 'batch_id': 'base_001', 'appended_rows': 216, 'total_rows': 216, 'version': 0}


{'feed': 'trips', 'batch_id': 'replay_002', 'appended_rows': 72, 'total_rows': 144, 'version': 1}


{
  "counts": {
    "trips": 144,
    "drivers": 6,
    "gps_events": 216
  },
  "checks": {
    "expected_totals": true,
    "two_trip_deliveries": true,
    "business_trip_count_unchanged": true,
    "initial_snapshot_retained": true,
    "base_csv_columns_are_strings": true,
    "payloads_and_source_hashes_preserved": true,
    "real_delta_files_present": true
  }
}
Retained: outputs/day01_bronze_i4xhsr5r/reports/bronze.json


<table dir="ltr" width="100%"><thead><tr><th width="50%" dir="ltr" lang="en" align="left">English</th><th width="50%" dir="rtl" lang="ar" align="right">العربية</th></tr></thead><tbody><tr><td width="50%" dir="ltr" lang="en" align="left" valign="top"><h2>Checks and next step</h2><p>A successful scenario has 144 Bronze trip deliveries but 72 distinct trip ids. Read <a href="../../day01/PRACTICE.md">the reasoning prompts</a>, complete Lab 01 notes, then use the same successful workspace in <a href="04_spark_scan.ipynb">Lab 02</a>. No generated table is automatically committed to Git.</p></td><td width="50%" dir="rtl" lang="ar" align="right" valign="top"><h2>الفحوص والخطوة التالية</h2><p>يحتوي السيناريو الناجح 144 سجل وصول في Bronze و72 معرف رحلة فريدًا. اقرأ <a href="../../day01/PRACTICE.md">أسئلة التفكير</a> وأكمل ملاحظات اللاب 01، ثم استخدم مساحة العمل الناجحة نفسها في <a href="04_spark_scan.ipynb">اللاب 02</a>. لا يضاف أي جدول مولد إلى Git تلقائيًا.</p></td></tr></tbody></table>